libevm/lib/evm.go

```diff
@@ -235,12 +275,26 @@ func (s *Service) EvmApply(params EvmParams) (error, *EvmResult) {
 		if traceErr != nil {
 			return fmt.Errorf("trace error: %v", traceErr), nil
 		}
-		// unfortunately, there is no way to get the tracer results without json marshaling, so we unmarshal again here
-		var traceResult *logger.ExecutionResult
-		if traceErr = json.Unmarshal(traceResultJson, &traceResult); traceErr != nil {
-			return fmt.Errorf("failed to unmarshal trace result %v", traceErr), nil
+
+		if params.TraceOptions.Tracer == nil { // Struct/opcode logger
+			var traceResult *logger.ExecutionResult
+			if traceErr = json.Unmarshal(traceResultJson, &traceResult); traceErr != nil {
+				return fmt.Errorf("failed to unmarshal trace result %v", traceErr), nil
+			}
+			result.TraceLogs = traceResult.StructLogs
+		} else if *params.TraceOptions.Tracer == "4byteTracer" { // 4byteTracer logger
+			var traceResult map[string]int
+			if traceErr = json.Unmarshal(traceResultJson, &traceResult); traceErr != nil {
+				return fmt.Errorf("failed to unmarshal trace result %v", traceErr), nil
+			}
+			result.FourByteTracerLogs = traceResult
+		} else if *params.TraceOptions.Tracer == "callTracer" { // callTracer logger
+			var traceResult *CallFrame
+			if traceErr = json.Unmarshal(traceResultJson, &traceResult); traceErr != nil {
+				return fmt.Errorf("failed to unmarshal trace result %v", traceErr), nil
+			}
+			result.CallTracerLogs = traceResult
```

Comment:
I think it would make sense to reduce overhead like this by just treating the in and output, i.e. tracer options and the results, as "raw" json.

My suggestion:
- Remove all the types for tracer results and their decoding and encoding. Just pass on the raw json from `tracer.GetResult()` and put it into our response to the RPC call. See `json.RawMessage` in GO and `@JsonRawValue` for the Jackson serializer on how to do that.
- Also remove the decoding and reencoding of tracer options. Just pass them on as raw json to the tracer. All of them have different options, which is why they do json parsing anyway. The only exception is the default StructLogger here.
- This is also how GETH does it to enable each tracer to have custom in and output formats without having to touch the other parts of the code.

Reply:
Agree on you point, I proceed to refactor the code as reported.
Now the input `tracerConfig` field and the tracer result response are managed are `json.RawMessage` in the Go part and as generic Object in the Scala/Java part. There was no need to flag them with `@JsonRawValue` annotation as they are correctly serialized as `LinkedHashMap` of `String` and `Object`.
Now we manage all the _geth_ native tracers and all its future changes in the interface (regarding the _geth_ version used) regardless of their interface.

The Go unit tests were modified just checking if the result is not null and that we don't receive any error from the `tracer.getResult()` method call. The checks on the response are now demanded to python tests where we interact with a live session of the application (test python `sc_evm_test_debug_methods.py`)

Reply:
Agree on all points, except maybe using `JsonNode` instead of `Object`. See comments below.